File: Top200/518-Change.java
package Top200;

class Solution {
    public int change(int amount, int[] coins) {
        // dp 数组定义，dp[i] 表示数量为 i 的种数
        int[] dp = new int[amount + 1];

        // dp 数组初始化
        dp[0] = 1;

        for (int i = 0; i < coins.length; i++) {
            for (int j = coins[i]; j <= amount; j++) {
                    dp[j] += dp[j - coins[i]];
            }
        }

        return dp[amount];
    }
}

File: Top200/518-Change.go
package main

func change(amount int, coins []int) int {
	l := len(coins)

	if l == 0 {
		return 0
	}

	dp := make([]int, amount+1)

	dp[0] = 1

	for i := 0; i < l; i++ {
		for j := coins[i]; j <= amount; j++ {
			dp[j] += dp[j-coins[i]]
		}
	}

	return dp[amount]
}
